### DISCUSSION TITLE
accessing `data` event from node `http`

### DISCUSSION THREAD

**Discussion by saihaj** [fastify/fastify#3388] [Q&A] [answered]
Consider this simple example
```js
const { createServer } = require('http')

const getBody = (req) => {
  let body = ''
  req.on('data', (chunk) => (body += chunk))
  return new Promise((resolve) =>
  req.on('end', () => resolve(JSON.parse(body))),
  )
}

const server = createServer(async (req, res) => {
    const a = await getBody(req)
    console.log(a)
    res.end(JSON.stringify({message:"should log"}))
})

server.listen(4000, () => {
  console.log(`Server is running at http://localhost:4000.`)
})
```

If I run that simple server and make a request I would expect `{ query: '{ hello }' }` to be logged in server and get `{"message":"should log"}` back in response.
```shell
curl -X POST \
-H "Content-Type: application/json" \
-d '{"query": "{ hello }"}' \
http://localhost:4000
```

According to docs https://www.fastify.io/docs/latest/Request/#request I can access incoming HTTP request from Node core using `request.raw` so consider this example:
```js
const fastify = require('fastify')
const server = fastify()

const getBody = (req) => {
  let body = ''
  req.on('data', (chunk) => (body += chunk))
  return new Promise((resolve) =>
  req.on('end', () => resolve(JSON.parse(body))),
  )
}

server.route({
  method: ['POST'],
  url: '/',
  async handler(req, res) {
    const a = await getBody(req.raw)
    console.log(a)
    res.send({ message: 'should log' })
  },
})

server.listen(4000, () => {
  console.log(`Server is running at http://localhost:4000.`)
})
```

If I try to make a request to this I don't get the same result. Not sure what I am doing wrong : )

**Comment by liuhanqu** [comment, marked as answer]
you can  see [this](https://github.com/fastify/fastify/blob/main/lib/contentTypeParser.js#L197)
i think the data had been consumed

**Comment by mcollina** [comment]
You do not need to do any of that as the body is already read for you in `req.body`. I noted this piece of information is not easily accessible / covered in the getting started, we should improve things!

**Reply by lauhon** [reply]
I just experimented and tried both things

1. req.body 
``` typescript
server.post ("/", (req, rep) => {
    logger.info({ body: req.body }, "My Request");
    rep.send("Done");
}
```

2. Data event
```typescript
server.post ("/", (req, rep) => {

    let body = "";

    req.raw.on("data", (data) => {
        body += data;
    });

    req.raw.on("end", () => {
       logger.info({body}, "My Request");
       rep.send("Done");
    });
}
```

Funny thing is, the second example never prints "My Request" and the Server never sends a response.
In my mind this means the following:

1. In Fastify you will never need the `data` event because its content is already there for you in `request.body`
2. Even if you want it is not possible to access the `data` event in Fastify, because Fastify already "exhausted" it in order to fill the `request.body` field

Are those assumptions correct or naw?

**Reply by lauhon** [reply]
Oh, I guess that's exactly what the accepted Answered by [liuhanqu](https://github.com/liuhanqu) means.  Just wasted my time on trying this 😆
